scripts/extract_features.py

```diff
@@ -0,0 +1,107 @@
+from __future__ import print_function, division
+
+import argparse
+
+import numpy as np
+import torch
+from torch.utils.data import Dataset
+from torchvision import transforms
+from torch import multiprocessing
+
+from goggle_classifier import load_data, train_model
+from Embedded2.src.jetson.main import FaceDetector
+
+VAL_SPLIT = .2
+activation = {}
+
+"""Save activation maps of images run partway through a CNN. Currently specific to Blazeface. 
+Use this script before running learn_features.py"""
+
+"""Transformations to perform on images before running inference. Blazeface requires 128x128 images."""
+data_transforms = {
+    'train': transforms.Compose([
+        transforms.Resize((128, 128)),
+        transforms.ToTensor()
+    ]),
+    'val': transforms.Compose([
+        transforms.Resize((128, 128)),
+        transforms.ToTensor()
+    ]),
+}
+
+
+# custom dataset for applying different transforms to train and val data
+class MapDataset(torch.utils.data.Dataset):
+    def __init__(self, dataset, map_fn):
+        self.dataset = dataset
+        self.map = map_fn
+
+    def __getitem__(self, item):
+        return self.map(self.dataset[item][0]), self.dataset[item][1]
+
+    def __len__(self):
+        return len(self.dataset)
+
+
+def run_inference(detector, data_location, data_loaders,  class_names):
+    """
+    Run all images in the dataset through the face detector, saving activation maps as .pt files.
+    @param detector: The FaceDetector. Must use a Blazeface model for now.
+    @param data_location: Directory in Imagefolder structure containing the images to train on.
+    @param data_loaders: A dictionary containing a 'train' DataLoader
+    and a 'val' DataLoader (returned by load_data).
+    @param class_names: List of class names. Used for saving .pt files in the same folder as their respective image.
+    """
+
+    i = 0
+    for phase in ['train', 'val']:
+        # iterate over all data
+        for inputs, labels in data_loaders[phase]:
+            inputs = inputs.to(device)
+            labels = labels.to(device)
+
+            # TODO blazeface never appears to detect any faces...
+            # the activation map is saved when .detect is called, so we need to call this even though it's not used
+            output = detector.detect(inputs)
+
+            # TODO I think there's a nicer way to write this file name, something like os.path.join?
+            # preferably should/could have the same name as the original image, but with a .pt extension.
+            # does the name get ignored when loading though?
+            torch.save(activation['backbone2'],
+                       data_location + '/' + class_names[labels.item()] + '/act_map' + str(i) + '.pt')
+            i += 1
+
+
+def get_activation(name):
+    # TODO find reference for why this is the correct way
+    def hook(model, input, output):
+        activation[name] = output.detach()
+
+    return hook
```

Comment:
based on this functions usage it looks like the developers wanted to give you more control over what is happening here. notice how here we are returning a function so that the program we pass it to can call it whenever. Maybe add some print statements in hook to see this in action, it might help clear things up.